engineering: add cli conventions

smola opened the pull request:
Signed-off-by: Santiago M. Mola <santi@mola.io>

smacker commented on engineering/conventions.md:
```diff
@@ -53,4 +54,13 @@ Refer to the [licensing](licensing.md), [maintainers](maintainers.md) and [docum
+
+## CLI
+
+Your application will expose one ore more binaries with a CLI.
```
> one ore more
one or more

ajnavarro commented on engineering/conventions-go.md:
```diff
@@ -62,4 +57,10 @@ A blank line is usually added after the end of any control structure, too.
+
+* Put sources for your executable commands in `cmd/<command-name>/main.go`.
+* We use [go-flags](https://github.com/jessevdk/go-flags) extensively for CLI options parsing.
+* Implement commands in a subpackage ([example](https://github.com/src-d/gitbase/blob/master/cli/gitbase/cmd/server.go)).
```
bad example, it is using `cli` (there is a PR fixing it)

smola replied:
@ajnavarro Ok. Once the PR is merged I'll update the link. Or maybe there is a better example?

ajnavarro commented on engineering/conventions-go.md:
```diff
@@ -62,4 +57,10 @@ A blank line is usually added after the end of any control structure, too.
+
+## CLI
+
+* Put sources for your executable commands in `cmd/<command-name>/main.go`.
```
Do you mean with this line that all the code must be inside `main.go` or maybe `cmd/<command-name>/main.go` should be `cmd/<command-name>/` ?

mcuadros commented:
Will be nice, include a full example. Also explain when short flags should be used, and when long ones, etc.

bzz commented:
👍 

It would be also nice to have a simple guideline for common, good-to-have CLI args i.e `-v/--version` and `-h/--help`

campoy commented on engineering/conventions-go.md:
```diff
@@ -62,4 +57,10 @@ A blank line is usually added after the end of any control structure, too.
+
+## CLI
+
+* Put sources for your executable commands in `cmd/<command-name>/`, with the `main` function in `cmd/<command-name>/main.go`.
```
Even when there's only one executable?

See `pga` as an example of something done the other way around.
No strong opinion, just wondering whether we need to retroactively apply this rule (which I had totally missed before)

smola replied:
Yes, we do this even for single executables. I guess we don't really need to go wild on retroactivity. For example, [rovers](https://github.com/src-d/rovers/) does not follow this, it uses the top level directory for the `main`, which is what we used to do in the past. But we'll probably change it at some point.

campoy commented:
We should also mention conventions regarding environment variables access and definition.

bzz commented:
@campoy on env vars, do you think https://github.com/src-d/guide/pull/214 might be a better place to do that?

smola commented:
Updated.

@mcuadros

> Will be nice, include a full example.

Do you mean a full example in terms of UX (e.g. the output of `cmd --help` with example help, flags, etc)? Or Go code example?

> Also explain when short flags should be used, and when long ones, etc.

I have clarified this: long flags should always be present, which is usually more user friendly when understanding calls, using autocompletion, etc. Short flags can be defined (in addition to long ones, not as an alternative), specially for cases where frequent direct calls by user are expected (e.g. `src-d/ml`, things like Babelfish' command line tools, etc). I would rather leave the decision on when to define the short ones open to each maintainer judgement though.

@bzz 

> It would also be nice to have a simple guideline for maintainers on exected common, good-to-have CLI args i.e -v/--version and -h/--help

Updated. `--help` and `--version` are probably the only ones that we should have globally (by the way, turns out this is the [GNU recommendation](https://www.gnu.org/prep/standards/html_node/Command_002dLine-Interfaces.html) too). For other cases, I just added that, when it makes sense, they should match wide-accepted behaviours (e.g. if you have an output file, you define `--output`, not `--target-file` or anything else).

@campoy 

> We should also mention conventions regarding environment variables access and definition.

Intentionally left out, since this is covered by https://github.com/src-d/guide/pull/214

dpordomingo commented on engineering/conventions-go.md:
```diff
@@ -69,3 +64,12 @@ A blank line is usually added after the end of any control structure, too.
 ### Naming
 
 * Prefer full names (e.g. `Repository`, not `Repo`).
+<<<<<<< HEAD
```
leftover

dennwc commented on engineering/conventions-go.md:
```diff
@@ -69,3 +64,9 @@ A blank line is usually added after the end of any control structure, too.
+## CLI
+
+* Put sources for your executable commands in `cmd/<command-name>/`, with the `main` function in `cmd/<command-name>/main.go`.
+* We use [go-flags](https://github.com/jessevdk/go-flags) extensively for CLI options parsing.
```
I still think we should use something less magical. If some flag is not being set properly from CLI it's pretty hard to see why because everything is hidden in reflection code.

I don't have any alternative to propose right now (except `cobra`, which is far less convenient), so I will leave my concerns for now. Will raise them a bit later when I have time to show an alternative that is both convenient to use and easy to debug.

bzz replied:
https://github.com/ogier/pflag ?
